File: generator.py
#!python
import json
import os
from urllib.request import urlopen
from random import choices as sample
from random import seed

if 'words.json' in os.listdir():
    with open('words.json', 'r') as f:
        words = json.load(f)
else:
    with urlopen('https://raw.githubusercontent.com/awong234/job-title-generator/main/words.json') as f:
        text = f.read().decode('utf-8')
    words = json.loads(text)

def generate_job_title(choices = None):
    if choices is None:
        choices = ["prefix", "adjective", "superlative", "job type", "title", "suffix"]
        # Sample from the space of choices, turn on and off randomly
        selection = sample([True, False], k = len(choices))
        selection = {x: y for x,y in zip(choices, selection)}
        # But keep job type active
        selection['job type'] = True
        choices = [ x for x,y in selection.items() if y]

    # Pick randomly from each category
    my_sample = [sample(words[x])[0] for x in choices]
    my_sample = ' '.join(my_sample)
    rank = sample([x for x in range(1,30)])[0]
    rank = int_to_roman(rank)
    my_sample = my_sample + ' ' + rank
    return my_sample

ROMAN = [
    (1000, "M"),
    ( 900, "CM"),
    ( 500, "D"),
    ( 400, "CD"),
    ( 100, "C"),
    (  90, "XC"),
    (  50, "L"),
    (  40, "XL"),
    (  10, "X"),
    (   9, "IX"),
    (   5, "V"),
    (   4, "IV"),
    (   1, "I"),
]

def int_to_roman(number):
    result = ""
    for (arabic, roman) in ROMAN:
        (factor, number) = divmod(number, arabic)
        result += roman * factor
    return result

print(generate_job_title())

# titles = []
# for i in range(0, 1000):
#     seed(i)
#     titles.append(generate_job_title())
#     print(i, titles[i])

# titles_sorted = titles.copy()
# titles_sorted.sort(key = lambda x: len(x))
# titles_sorted

File: generator.R
#!Rscript
library(jsonlite)

if (file.exists('words.json')) {
    words <- jsonlite::fromJSON("words.json")
} else {
    words <- jsonlite::fromJSON("https://raw.githubusercontent.com/awong234/job-title-generator/main/words.json")
}

generate_job_title = function(choices = NULL) {
    if (is.null(choices)) {
        choices = c("prefix", "adjective", "superlative", "job type", "title", "suffix")
        # Sample from the space of choices, turn on and off randomly
        selection = sample(x = c(TRUE, FALSE), size = length(choices), replace = TRUE)
        names(selection) = choices
        # But keep job type active
        selection['job type'] = TRUE
        choices = choices[selection]
    }
    # Pick randomly from each category
    my_sample = sapply(X = choices, FUN = \(x) sample(words[[x]], size = 1))
    rank = sample(c(1:29), size = 1) |> as.roman()
    out = paste0( my_sample , collapse = ' ')
    out = paste(out, rank)
    out
}

cat(generate_job_title(), '\n')

# titles = vector(mode = 'character', length = 1000)
# for (i in 1:1000) {
#     set.seed(i)
#     titles[i] = generate_job_title()
#     cat(i, '\t', titles[i], '\n')
# }

# titles_df = data.frame(
#     seed = 1:1000,
#     titles
# )

# Titles sorted by length
# titles_df[order(titles_df$titles |> nchar()), ]
